plugins/net_plugin/net_plugin.cpp

```diff
@@ -1244,24 +1251,78 @@ namespace eosio {
+      return create_send_buffer( signed_block_which, *sb );
+   }
+
+   static std::shared_ptr<std::vector<char>> create_send_buffer( const signed_block_v0& sb_v0 ) {
+      static_assert( signed_block_v0_which == net_message::position<signed_block_v0>() );
+      // this implementation is to avoid copy of signed_block_v0 to net_message
+      // matches which of net_message for signed_block_v0
+      fc_dlog( logger, "sending v0 block ${bn}", ("bn", sb_v0.block_num()) );
+      return create_send_buffer( signed_block_v0_which, sb_v0 );
+   }
+
+   static std::shared_ptr<std::vector<char>> create_send_buffer( const packed_transaction_ptr& trx ) {
+      static_assert( trx_message_v1_which == net_message::position<trx_message_v1>() );
+      fc::optional<transaction_id_type> trx_id;
+      if( trx->get_estimated_size() > 1024 ) { // simple guess on threshold
+         fc_dlog( logger, "including trx id, est size: ${es}", ("es", trx->get_estimated_size()) );
+         trx_id = trx->id();
+      }
+      // const cast required, trx_message_v1 has non-const shared_ptr because FC_REFLECT does not work with const types
+      trx_message_v1 v1{ std::move(trx_id), std::const_pointer_cast<packed_transaction>(trx) };
+      return create_send_buffer( trx_message_v1_which, v1 );
+   }
+
+   static std::shared_ptr<std::vector<char>> create_send_buffer( const packed_transaction_v0& trx ) {
+      static_assert( packed_transaction_v0_which == net_message::position<packed_transaction_v0>() );
+      // this implementation is to avoid copy of packed_transaction_v0 to net_message
+      // matches which of net_message for packed_transaction_v0
+      return create_send_buffer( packed_transaction_v0_which, trx );
    }
 
-   static std::shared_ptr<std::vector<char>> create_send_buffer( const packed_transaction& trx ) {
-      // this implementation is to avoid copy of packed_transaction to net_message
-      // matches which of net_message for packed_transaction
-      auto pt_v0 = trx.to_packed_transaction_v0();
-      return create_send_buffer( packed_transaction_v0_which, *pt_v0 );
+   static std::shared_ptr<std::vector<char>> get_send_buffer( connection& c,
+                                                              const signed_block_ptr& b,
+                                                              std::shared_ptr<std::vector<char>>& send_buffer,
+                                                              std::shared_ptr<std::vector<char>>& send_buffer_v0,
+                                                              bool& valid_v0_block )
+   {
+      const uint16_t v = c.protocol_version.load();
+      if( v >= proto_pruned_types ) {
+         if( !send_buffer ) {
+            send_buffer = create_send_buffer( b );
+         }
+         return send_buffer;
+      } else {
+         if( !send_buffer_v0 ) {
+            const auto sb_v0 = valid_v0_block ? b->to_signed_block_v0() : signed_block_v0_uptr();
```

Comment:
valid_v0_block is initialized at the start and is really just pulling the check in to_signed_block_v0() out and making 2 sets of logic to maintain instead of 1.  Just initialize valid_v0_block to true and let it get set to false by this logic.